modules/airavata-metascheduler/process-scheduler/src/main/java/org/apache/airavata/metascheduler/process/scheduling/engine/cr/selection/ComputeResourceSelectionPolicyImpl.java

```diff
@@ -0,0 +1,108 @@
+package org.apache.airavata.metascheduler.process.scheduling.engine.cr.selection;
+
+import org.apache.airavata.common.utils.ThriftClientPool;
+import org.apache.airavata.metascheduler.core.engine.ComputeResourceSelectionPolicy;
+import org.apache.airavata.metascheduler.core.utils.Utils;
+import org.apache.airavata.model.appcatalog.groupresourceprofile.GroupComputeResourcePreference;
+import org.apache.airavata.model.appcatalog.groupresourceprofile.GroupResourceProfile;
+import org.apache.airavata.model.appcatalog.userresourceprofile.UserComputeResourcePreference;
+import org.apache.airavata.model.appcatalog.userresourceprofile.UserResourceProfile;
+
+import org.apache.airavata.registry.api.RegistryService;
+import org.apache.airavata.registry.api.RegistryService.Client;
+import org.apache.airavata.registry.api.exception.RegistryServiceException;
+
+public abstract class ComputeResourceSelectionPolicyImpl implements ComputeResourceSelectionPolicy {
+
+    protected ThriftClientPool<RegistryService.Client> registryClientPool;
+
+    public ComputeResourceSelectionPolicyImpl() {
+        this.registryClientPool =Utils.getRegistryServiceClientPool();
+    }
+
+    private boolean isValid(String str) {
+        return str != null && !str.trim().isEmpty();
+    }
+
+    public UserResourceProfile getUserResourceProfile(String username, String gatewayId) throws Exception {
+        RegistryService.Client client = this.registryClientPool.getResource();
+        try {
+            return client.getUserResourceProfile(username, gatewayId);
+        }finally {
+            this.registryClientPool.returnResource(client);
+        }
+
+    }
+
+
+    private UserComputeResourcePreference getUserComputeResourcePreference(String gatewayId, String username,
+                                                                           String computeResourceId) throws Exception {
+        RegistryService.Client client = this.registryClientPool.getResource();
+        try {
+            return client.getUserComputeResourcePreference(username, gatewayId, computeResourceId);
+        }finally {
+            this.registryClientPool.returnResource(client);
+        }
+    }
+
+    public String getComputeResourceCredentialToken(String gatewayId,
+                                                    String username, String computeResourceId, boolean isUseUserCRPref,
+                                                    boolean isSetGroupResourceProfileId, String groupResourceProfileId) throws Exception {
+        if (isUseUserCRPref) {
+            if (getUserComputeResourcePreference(gatewayId, username, computeResourceId) != null &&
```

Comment:
getUserComputeResourcePreference will never return null because RegistryService.Client.getUserComputeResourcePreference() will never return null. Thrift RPC methods can't return null and instead will throw an exception.  In getUserComputeResourcePreference() I think you could catch the exception and return null, but ideally we'd have a Thrift method that does the existence check and returns a boolean (like RegistryServerHandler.isUserResourceProfileExists).

Reply:
likewise for 
- getUserResourceProfile
- getGroupComputeResourcePreference
- getGroupResourceProfile